misc/scripts/header_guards.sh

```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+
+if [ ! -f "version.py" ]; then
+  echo "Warning: This script is intended to be run from the root of the Godot repository."
+  echo "Some of the paths checks may not work as intended from a different folder."
+fi
+
+for file in $(find -name "thirdparty" -prune -o -name "*.h" -print); do
+  # Skip *.gen.h and *-so_wrap.h, they're generated.
+  if [[ "$file" == *".gen.h" || "$file" == *"-so_wrap.h" ]]; then continue; fi
+  # Has important define before normal header guards.
+  if [[ "$file" == *"thread.h" || "$file" == *"platform_config.h" ]]; then continue; fi
+
+  bname=$(basename $file .h)
+
+  # Add custom prefix or suffix for generic filenames with a well-defined namespace.
+
+  prefix=
+  if [[ "$file" == "./modules/"*"/register_types.h" ]]; then
+    module=$(echo $file | sed "s@.*modules/\([^/]*\).*@\1@")
+    prefix="${module^^}_"
+  fi
+  if [[ "$file" == "./platform/"*"/api/api.h" || "$file" == "./platform/"*"/export/"* ]]; then
+    platform=$(echo $file | sed "s@.*platform/\([^/]*\).*@\1@")
+    prefix="${platform^^}_"
+  fi
+  if [[ "$file" == "./modules/mono/utils/"* && "$bname" != *"mono"* ]]; then prefix="MONO_"; fi
+  if [[ "$file" == "./servers/rendering/storage/utilities.h" ]]; then prefix="RENDERER_"; fi
+
+  suffix=
+  if [[ "$file" == *"dummy"* && "$bname" != *"dummy"* ]]; then suffix="_DUMMY"; fi
+  if [[ "$file" == *"gles3"* && "$bname" != *"gles3"* ]]; then suffix="_GLES3"; fi
+  if [[ "$file" == *"renderer_rd"* && "$bname" != *"rd"* ]]; then suffix="_RD"; fi
```

Comment:
FYI @clayjohn @BastiaanOlij - this pretty much matches what was already done manually, though it adds `_RD` to a few files which weren't using it and might not have multiple implementations for the same filename.